### ISSUE TITLE
Variables known to be non-null sometimes artifically restricted to one variable

### ISSUE THREAD

**Issue by gafter** [dotnet/roslyn#33347]
In a number of places in the `NullableWalker`, we set a single variable to the non-null state.  But in many of those cases, we should be using `GetSlotsToMarkAsNotNullable()` to mark more than one variable non-null.

Here is an example of a test for one of them:

``` c#
        [Fact]
        public void NestedNullConditionalAccess()
        {
            var source =
@"class Node
{
    public Node? Next = null;
    void M(Node node) { }
    private static void Test(Node? node)
    {
        node?.Next?.Next?.M(node.Next);
    }
}
";
            var comp = CreateCompilation(new[] { source }, options: WithNonNullTypesTrue());
            comp.VerifyDiagnostics();
        }
```

this test fails - the compiler incorrectly reports

``` none

                // (7,29): warning CS8604: Possible null reference argument for parameter 'node' in 'void Node.M(Node node)'.
                //         node?.Next?.Next?.M(node.Next);
                Diagnostic(ErrorCode.WRN_NullReferenceArgument, "node.Next").WithArguments("node", "void Node.M(Node node)").WithLocation(7, 29)
```

Tests and comments in the code will refer to this issue.

**Comment by gafter** [comment]
Here is another example (though I'm not sure it is the same issue)
``` c#
#nullable enable
class Node
{
    public Node? Next = null;
    static Node M2(Node a, Node b) => a;
    Node M1() => null!;

    private static void Test(Node notNull, Node? possiblyNull)
    {
        _ = possiblyNull?.Next?.M1() ?? M2(possiblyNull = notNull, possiblyNull.Next = notNull);
        possiblyNull.Next.M1(); // incorrect warning
    }
}
```

**Comment by jcouv** [comment]
Relates to https://github.com/dotnet/roslyn/issues/31909 (possible duplicate even).

**Comment by jcouv** [comment]
I have a PR out that addresses these specific scenarios (the problem was that couldn't assign slots properly in conditional access, because the conditional receiver is a placeholder for the actual receiver which we visited).
For the more general question, I believe that we call `GetSlotsToMarkAsNotNullable()` adequately already (in places that do a null test).
